[SDK 50] expo-file-system - `uploadAsync` Error: Unable to upload the file (cancelled)

I am getting this error when we try to upgrade to Expo 50. Nothing I change seems to alter the effect. I investigated a lot of the commits in `expo-file-system` in the last few months and didn't see anything suspicious that caught my attention. I'm pretty stumped as to what could be happening here...

Here is what I see when I print the error thrown by `uploadAsync`:
```
[Error: Unable to upload the file: 'Error Domain=NSURLErrorDomain Code=-999 "cancelled" UserInfo={NSErrorFailingURLStringKey=https://api.cloudinary.com/v1_1/questmate-testing/auto/upload, NSErrorFailingURLKey=https://api.cloudinary.com/v1_1/questmate-testing/auto/upload, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "BackgroundUploadTask <8403A4A5-8290-4773-A91B-58832FF985F0>.<12>",
    "LocalUploadTask <8403A4A5-8290-4773-A91B-58832FF985F0>.<12>"
), _NSURLErrorFailingURLSessionTaskErrorKey=BackgroundUploadTask <8403A4A5-8290-4773-A91B-58832FF985F0>.<12>, NSLocalizedDescription=cancelled}']
```

If I stringify the error it prints:
```
{"code":"ERR_FILESYSTEM_CANNOT_UPLOAD"}
```

Our call to `uploadAsync` looks like the following (where parameters is a `Record<string, string>`):
```typescript
const uploadResult = await uploadAsync(url, mediaItem.uri, {
    httpMethod: "POST",
    sessionType: FileSystemSessionType.BACKGROUND,
    uploadType: FileSystemUploadType.MULTIPART,
    fieldName: "file",
    parameters,
});
```

We double checked the config we pass everything looks good. Plus we did not change any code in this area during the upgrade from Expo 49 to Expo 50.

We are using `"expo-file-system": "16.0.5"`


_Originally posted by @Jackman3005 in https://github.com/expo/expo/issues/26508#issuecomment-1913536491_

I tried downgrading to `"expo-file-system": "15.4.3"` which is the version we were using before upgrading to Expo 50. But it did not compile for the following reason. [Looks like references to `bundleAssets` were removed in `16.0.0`](https://github.com/expo/expo/blob/main/packages/expo-file-system/CHANGELOG.md#1600--2023-12-12) so we will try that version and see if there is any effect on the cancelled error.

```
Compiling expo-file-system Pods/EXFileSystem » NSData+EXFileSystem.m

› Compiling expo-file-system Pods/EXFileSystem » EXTaskHandlersManager.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionUploadTaskDelegate.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionTaskDispatcher.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionTaskDelegate.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionResumableDownloadTaskDelegate.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionHandler.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionDownloadTaskDelegate.m

› Compiling expo-file-system Pods/EXFileSystem » EXSessionCancelableUploadTaskDelegate.m

› Compiling expo-file-system Pods/EXFileSystem » EXFileSystemLocalFileHandler.m


⚠️  (../../../../../ios/EXFileSystem/EXFileSystem.h:23:49)

- (NSString *)generatePathInDirectory:(NSString *)directory withExtension:(NSString *)extension;
                                                ^ pointer is missing a nullability type specifier [-Wnullability-completeness]

› Compiling expo-file-system Pods/EXFileSystem » EXFileSystemAssetLibraryHandler.m

› Compiling expo-file-system Pods/EXFileSystem » EXFileSystem.m

› Compiling expo-file-system Pods/EXFileSystem » EXFileSystem-dummy.m

› Compiling expo-file-system Pods/EXFileSystem » EXFilePermissionModule.m


❌  (/Users/expo/workingdir/build/app/node_modules/expo-file-system/ios/EXFileSystem/EXFilePermissionModule.m:42:42)

  40 |   }
  41 | 
> 42 |   NSString *bundleDirectory = fileSystem.bundleDirectory;
     |                                          ^ property 'bundleDirectory' not found on object of type '__strong id<EXFileSystemInterface>'
  43 |   if (bundleDirectory != nil && [path hasPrefix:[bundleDirectory stringByAppendingString:@"/"]]) {
  44 |     return EXFileSystemPermissionRead;
  45 |   }

› Compiling expo-file-system Pods/EXFileSystem » NSData+EXFileSystem.m

› Compiling expo-file-system Pods/EXFileSystem » EXTaskHandlersManager.m

⚠️  Script has ambiguous dependencies causing it to run on every build.
   To fix, go to: Xcode » QuestmateDev/QuestmateDev » Build Phases » 'Upload Debug Symbols to Sentry'
   Either: Uncheck "Based on dependency analysis", or select output files to trigger the script

▸ ** BUILD FAILED **

▸ The following build commands failed:

▸ 	CompileC /Users/expo/workingdir/build/app/ios/build/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/EXFileSystem.build/Objects-normal/x86_64/EXFilePermissionModule.o /Users/expo/workingdir/build/app/node_modules/expo-file-system/ios/EXFileSystem/EXFilePermissionModule.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler (in target 'EXFileSystem' from project 'Pods')

▸ (1 failure)

** BUILD FAILED **

The following build commands failed:

	CompileC /Users/expo/workingdir/build/app/ios/build/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/EXFileSystem.build/Objects-normal/x86_64/EXFilePermissionModule.o /Users/expo/workingdir/build/app/node_modules/expo-file-system/ios/EXFileSystem/EXFilePermissionModule.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler (in target 'EXFileSystem' from project 'Pods')

(1 failure)

Exit status: 65
```

--- a/packages/expo-file-system/CHANGELOG.md
+++ b/packages/expo-file-system/CHANGELOG.md
@@ -9,7 +9,7 @@
 ### 🐛 Bug fixes
 
 - On `iOS`, set `httpMethod` on upload requests. ([#26516](https://github.com/expo/expo/pull/26516) by [@alanjhughes](https://github.com/alanjhughes))
-- On `iOS`, fix upload task requests.
+- On `iOS`, fix upload task requests. ([#26880](https://github.com/expo/expo/pull/26880) by [@alanjhughes](https://github.com/alanjhughes))
 
 ### 💡 Others
 

--- a/packages/expo-file-system/CHANGELOG.md
+++ b/packages/expo-file-system/CHANGELOG.md
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - On `iOS`, set `httpMethod` on upload requests. ([#26516](https://github.com/expo/expo/pull/26516) by [@alanjhughes](https://github.com/alanjhughes))
+- On `iOS`, fix upload task requests.
 
 ### 💡 Others
 

--- a/packages/expo-file-system/ios/FileSystemExceptions.swift
+++ b/packages/expo-file-system/ios/FileSystemExceptions.swift
@@ -67,3 +67,15 @@ final class CannotDetermineDiskCapacity: Exception {
     "Unable to determine free disk storage capacity"
   }
 }
+
+final class FailedToCreateBodyException: Exception {
+  override var reason: String {
+    "Unable to create multipart body"
+  }
+}
+
+final class FailedToAccessDirectoryException: Exception {
+  override var reason: String {
+    "Failed to access `Caches` directory"
+  }
+}

--- a/packages/expo-file-system/ios/FileSystemModule.swift
+++ b/packages/expo-file-system/ios/FileSystemModule.swift
@@ -156,7 +156,7 @@ public final class FileSystemModule: Module {
         throw FileNotExistsException(localUrl.path)
       }
       let session = options.sessionType == .background ? backgroundSession : foregroundSession
-      let task = createUploadTask(session: session, targetUrl: targetUrl, sourceUrl: localUrl, options: options)
+      let task = try createUploadTask(session: session, targetUrl: targetUrl, sourceUrl: localUrl, options: options)
       let taskDelegate = EXSessionUploadTaskDelegate(resolve: promise.resolver, reject: promise.legacyRejecter)
 
       sessionTaskDispatcher.register(taskDelegate, for: task)
@@ -165,7 +165,7 @@ public final class FileSystemModule: Module {
 
     AsyncFunction("uploadTaskStartAsync") { (targetUrl: URL, localUrl: URL, uuid: String, options: UploadOptions, promise: Promise) in
       let session = options.sessionType == .background ? backgroundSession : foregroundSession
-      let task = createUploadTask(session: session, targetUrl: targetUrl, sourceUrl: localUrl, options: options)
+      let task = try createUploadTask(session: session, targetUrl: targetUrl, sourceUrl: localUrl, options: options)
       let onSend: EXUploadDelegateOnSendCallback = { [weak self] _, _, totalBytesSent, totalBytesExpectedToSend in
         self?.sendEvent(EVENT_UPLOAD_PROGRESS, [
           "uuid": uuid,

--- a/packages/expo-file-system/ios/NetworkingHelpers.swift
+++ b/packages/expo-file-system/ios/NetworkingHelpers.swift
@@ -30,7 +30,7 @@ func createUrlRequest(url: URL, headers: [String: String]?) -> URLRequest {
   return request
 }
 
-func createUploadTask(session: URLSession, targetUrl: URL, sourceUrl: URL, options: UploadOptions) -> URLSessionUploadTask {
+func createUploadTask(session: URLSession, targetUrl: URL, sourceUrl: URL, options: UploadOptions) throws -> URLSessionUploadTask {
   var request = createUrlRequest(url: targetUrl, headers: options.headers)
   request.httpMethod = options.httpMethod.rawValue
 
@@ -39,47 +39,60 @@ func createUploadTask(session: URLSession, targetUrl: URL, sourceUrl: URL, optio
     return session.uploadTask(with: request, fromFile: sourceUrl)
   case .multipart:
     let boundaryString = UUID().uuidString
-    let data = try? createMultipartBody(boundary: boundaryString, sourceUrl: sourceUrl, options: options)
+    guard let data = createMultipartBody(boundary: boundaryString, sourceUrl: sourceUrl, options: options) else {
+      throw FailedToCreateBodyException()
+    }
 
     request.setValue("multipart/form-data; boundary=\(boundaryString)", forHTTPHeaderField: "Content-Type")
-    request.httpBody = data
 
-    return session.uploadTask(withStreamedRequest: request)
+    let localURL = try createLocalUrl(from: sourceUrl)
+    try? data.write(to: localURL)
+
+    return session.uploadTask(with: request, fromFile: localURL)
   }
 }
 
-func createMultipartBody(boundary: String, sourceUrl: URL, options: UploadOptions) throws -> Data {
-  let fileName = options.fieldName ?? sourceUrl.lastPathComponent
-  let fileContents = try String(contentsOf: sourceUrl)
-  let mimeType = options.mimeType ?? findMimeType(forAttachment: sourceUrl)
-
-  let body = """
-\(headersForMultipartParams(options.parameters, boundary: boundary))
---\(boundary)
-Content-Disposition: form-data; name="\(fileName)"; filename="\(sourceUrl.lastPathComponent)"
-Content-Type: \(mimeType)
-
-\(fileContents)
---\(boundary)--
-"""
-
-  guard let bodyData = body.data(using: .utf8) else {
-    throw HeaderEncodingFailedException(sourceUrl.absoluteString)
+func createLocalUrl(from sourceUrl: URL) throws -> URL {
+  guard let cachesDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+    throw FailedToAccessDirectoryException()
   }
-  return bodyData
+  let tempDir = cachesDir.appendingPathComponent("uploads")
+  FileSystemUtilities.ensureDirExists(at: tempDir)
+  return tempDir.appendingPathComponent(sourceUrl.lastPathComponent)
 }
 
-func headersForMultipartParams(_ params: [String: String]?, boundary: String) -> String {
-  guard let params else {
-    return ""
+func createMultipartBody(boundary: String, sourceUrl: URL, options: UploadOptions) -> Data? {
+  let fieldName = options.fieldName ?? sourceUrl.lastPathComponent
+  var mimeType = options.mimeType ?? findMimeType(forAttachment: sourceUrl)
+  guard let data = try? Data(contentsOf: sourceUrl) else {
+    return nil
   }
-  return params.map { (key: String, value: String) in
-"""
---\(boundary)
-Content-Disposition: form-data; name="\(key)"
 
-\(value)
-"""
+  var body = Data()
+  headersForMultipartParams(options.parameters, boundary: boundary, body: &body)
+
+  body.append("--\(boundary)\r\n".data)
+  body.append("Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(sourceUrl.lastPathComponent)\"\r\n".data)
+  body.append("Content-Type: \(mimeType)\r\n\r\n".data)
+  body.append(data)
+  body.append("\r\n".data)
+  body.append("--\(boundary)--\r\n".data)
+
+  return body
+}
+
+func headersForMultipartParams(_ params: [String: String]?, boundary: String, body: inout Data) {
+  if let params {
+    for (key, value) in params {
+      body.append("--\(boundary)\r\n".data)
+      body.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n".data)
+      body.append("\(value)\r\n".data)
+    }
   }
-  .joined()
+}
+
+// All swift strings are unicode correct.
+// This avoids the optional created by string.data(using: .utf8)
+private extension String {
+  var data: Data { Data(self.utf8) }
 }

--- a/packages/expo-file-system/ios/Tests/EXFileSystemSpec.swift
+++ b/packages/expo-file-system/ios/Tests/EXFileSystemSpec.swift
@@ -20,7 +20,7 @@ class EXFileSystemSpec: ExpoSpec {
       it("should handle UTF-8 characters") {
         let utf8UriInput = "file:///var/mobile/中文"
         let utf8UriExpectedOutput = "file:///var/mobile/%E4%B8%AD%E6%96%87"
-        let utf8Uri = fileSystem.percentEncodedURL(fromURIString:utf8UriInput)
+        let utf8Uri = fileSystem.percentEncodedURL(fromURIString: utf8UriInput)
 
         expect(utf8Uri?.absoluteString) == utf8UriExpectedOutput
         expect(utf8Uri?.scheme) == "file"
@@ -29,7 +29,7 @@ class EXFileSystemSpec: ExpoSpec {
       it("should handle URI with percent, numbers and UTF-8 characters") {
         let input = "file:///document/directory/%40%2F中文"
         let expectedOutput = "file:///document/directory/%40%2F%E4%B8%AD%E6%96%87"
-        let uri = fileSystem.percentEncodedURL(fromURIString:input)
+        let uri = fileSystem.percentEncodedURL(fromURIString: input)
 
         expect(uri?.absoluteString) == expectedOutput
       }
@@ -37,7 +37,7 @@ class EXFileSystemSpec: ExpoSpec {
       it("should not decode percentages in URI") {
         let input = "file:///document/hello%2Fworld.txt"
         let unexpectedOutput = "file:///document/hello/world.txt"
-        let uri = fileSystem.percentEncodedURL(fromURIString:input)
+        let uri = fileSystem.percentEncodedURL(fromURIString: input)
 
         // Should not create a directory named "hello"
         expect(uri?.absoluteString) != unexpectedOutput
@@ -45,7 +45,7 @@ class EXFileSystemSpec: ExpoSpec {
 
       it("should handle assets-library URIs") {
         let assetsLibraryUriInput = "assets-library://asset/asset.JPG?id=3C1D9C54-9521-488F-BB27-AA1EA0F8AF04/L0/001&ext=JPG"
-        let assetsLibraryUri = fileSystem.percentEncodedURL(fromURIString:assetsLibraryUriInput)
+        let assetsLibraryUri = fileSystem.percentEncodedURL(fromURIString: assetsLibraryUriInput)
 
         expect(assetsLibraryUri?.absoluteString) == assetsLibraryUriInput
         expect(assetsLibraryUri?.scheme) == "assets-library"

